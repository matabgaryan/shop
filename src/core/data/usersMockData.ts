import {Role} from "../enums";
import {Pojo} from "../types";

export const usersMockData: Pojo = {
    user: {
        id: '1',
        username: 'user',
        email: 'user@exampl.com',
        firstName: 'John',
        lastName: 'Doe',
        roles: [Role.USER],
    },
    admin: {
        id: '2',
        username: 'admin',
        email: 'admin@exampl.com',
        firstName: 'admin',
        lastName: 'Doe',
        roles: [Role.ADMIN, Role.USER],
    },
}
