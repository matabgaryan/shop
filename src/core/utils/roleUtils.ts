import {Role} from "../enums";

export const checkUserRole = (routeRoles: Role[], authenticatedUserRoles: Role[]) =>  {
    return routeRoles.some(item => authenticatedUserRoles.includes(item))
}
