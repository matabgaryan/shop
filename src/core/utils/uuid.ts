import { v4 as uuidv4 } from "uuid";

export const generatedUuid: () => string = () => {
    return uuidv4();
}
