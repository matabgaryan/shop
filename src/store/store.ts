import {combineReducers, configureStore, PreloadedState} from "@reduxjs/toolkit";
import shopCart from "./slices/shopCart";
import productsReducer from "./slices/products";
import authReducer from './slices/auth';
import modalReducer from './slices/modals';
import {RootState} from "../core/types";

export const store = configureStore({
    reducer:{
        shopCart,
        products: productsReducer,
        auth: authReducer,
        modals: modalReducer,

    },
    devTools: true,
})
const rootReducer = combineReducers({
    shopCart,
    products: productsReducer,
    auth: authReducer,
    modals: modalReducer,
});


export function setupStore(preloadedState?: PreloadedState<RootState>) {
    return configureStore({
        reducer: rootReducer,
        preloadedState
    })
}

