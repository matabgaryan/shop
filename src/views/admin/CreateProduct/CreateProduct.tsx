import React  from 'react'
import {FormInput} from "../../../components/inputs/FormInput";
import {Container} from "../../../components/container/Container";
import {useCreateProductHook} from "../../../hooks/useCreateProduct";


export const CreateProduct:React.FC = ()  => {
    const { formFields,handleFormChange, handleCancel, submit, addFields, removeFields}= useCreateProductHook();

    return (
        <Container>
        <div className="mt-16 mx-auto max-w-4xl w-full flex flex-col items-center">
            <p className="mb-8 font-GoogleSansBold text-[18px] text-appBlack font-[700]">Configure Product</p>
            <form onSubmit={submit}>
                {
                    formFields.map((item: any, index: number) => {
                        return(
                            <div
                                key={index}
                                className="grid gap-4  mb-6 lg:grid-cols-3 mb-6 rounded-sm border-b border-lightGrey	pb-4">
                                <FormInput
                                    name="name"
                                    labelName="Name"
                                    placeholder="Specify product name"
                                    onChange={event => handleFormChange(event, index)}
                                />
                                <FormInput
                                    name="price"
                                    labelName="Price"
                                    placeholder="Specify product price"
                                    onChange={event => handleFormChange(event, index)}
                                />
                                <FormInput
                                    name="weight"
                                    labelName="Weight"
                                    placeholder="Weight"
                                    onChange={event => handleFormChange(event, index)}
                                />
                                <FormInput
                                    name="startDate"
                                    labelName="Start Date"
                                    placeholder="Start Date"
                                    onChange={event => handleFormChange(event, index)}
                                />
                                <FormInput
                                    name="endDate"
                                    labelName="End Date"
                                    placeholder="End Date"
                                    onChange={event => handleFormChange(event, index)}
                                />
                                <div className="flex flex-row  items-start justify-center">
                                    <button
                                        onClick={() => removeFields(index)}
                                        className=" text-white bg-oziRed hover:opacity-70  font-medium rounded-lg text-sm  px-5 py-2.5 text-center ">
                                        Remove
                                    </button>
                                </div>
                            </div>
                        )
                    })
                }

            <div>
                <div className="flex items-start justify-center">
                    <button
                        onClick={addFields}
                        className="text-appBlack text-[16px] font-GoogleSansRegular font-[400] focus:outline-none  text-center">
                        + Add more
                    </button>
                </div>
            </div>
            <div className="flex justify-end mb-8 mt-5">
                <button
                    onClick={() => handleCancel()}
                    className="mr-4 w-150 text-appBlack bg-transparent hover:opacity-80	 focus:ring-4 font-medium rounded-lg text-sm  px-5 py-2.5 text-center ">
                    Cancel
                </button>
                <button
                    type="submit"
                    onClick={submit}
                    className="w-150 text-white bg-appBlack hover:opacity-70  font-medium rounded-lg text-sm  px-5 py-2.5 text-center ">
                    Add
                </button>
            </div>
        </form>
        </div>
        </Container>
    )
}
