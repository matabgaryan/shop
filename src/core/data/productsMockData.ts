import {generatedUuid} from "../utils/uuid";
import {IProduct} from "../../store/slices/interface";

export const productsData: IProduct[] = [
    {
        id: generatedUuid(),
        name: 'Product 1',
        price: '30',
        weight: 100,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 2',
        price: '30',
        weight: 100,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 3',
        price: '30',
        weight: 20,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 4',
        price: '30',
        weight: 30,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 5',
        price: '30',
        weight: 40,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 6',
        price: '30',
        weight: 200,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 7',
        price: '30',
        weight: 50,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 8',
        price: '30',
        weight: 100,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 9',
        price: '30',
        weight: 65,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 10',
        price: '30',
        weight: 70,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
]
