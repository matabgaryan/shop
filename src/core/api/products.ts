import { productsData } from '../data/productsMockData'
import {IProduct} from "../../store/slices/interface";

export const getProductsList: () => Promise<IProduct[]> = () => {
    return  new Promise((resolve) => {
        setTimeout(() => resolve(productsData), 250);
    });
}
