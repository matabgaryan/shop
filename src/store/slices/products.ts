import {ActionReducerMapBuilder, createAsyncThunk, createSlice, SliceCaseReducers} from "@reduxjs/toolkit";
import {STATUS} from "../../core/enums";
import {getProductsList} from "../../core/api/products";
import {IProduct, IProductsState} from "./interface";
import {NoInfer} from "@reduxjs/toolkit/dist/tsHelpers";
import {RootState} from "../../core/types";

const initialState: IProductsState = {
    status: "",
    products: [],
};

const productSlice = createSlice<IProductsState, SliceCaseReducers<IProductsState>>({
    name: "products",
    initialState,
    reducers: {
        addProducts: (state, action) => {
            state.products.push(...action.payload);
        }
    },
    extraReducers: (builder: ActionReducerMapBuilder<NoInfer<IProductsState>>) => {
        builder
            .addCase(getProducts.pending, (state) => {
                state.status = STATUS.PENDING;
            })
            .addCase(getProducts.fulfilled, (state, action) => {
                state.products = action.payload;
                state.status = STATUS.SUCCEEDED;
            })
            .addCase(getProducts.rejected, (state) => {
                state.status = STATUS.REJECTED;
            });
    },
});

export const getProducts = createAsyncThunk<IProduct[]>("fetch/products", async () => {
    return await getProductsList();
});
export const { addProducts } = productSlice.actions;

export const filteredProducts = (state: RootState) => state.products.products

export default productSlice.reducer;
