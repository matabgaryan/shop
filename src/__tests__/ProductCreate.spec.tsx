import {
    render,
    screen,
    fireEvent,
    RenderResult,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import {  MemoryRouter } from 'react-router-dom';
import {setupStore, store} from '../store/store';
import {renderWithProviders} from "../core/utils/testUtils";
import {productsData} from "../core/data/productsMockData";
import { addProducts } from "../store/slices/products";
import {CreateProduct} from "../views/admin/CreateProduct/CreateProduct";
const renderCreateProduct = (): RenderResult =>
    render(
        <Provider store={store}>
            <MemoryRouter>
                <CreateProduct  />
            </MemoryRouter>
        </Provider>
    );
describe('Create Products', () => {
    test('Take Snapshot', () => {
        const { asFragment } = renderCreateProduct();
        expect(asFragment()).toMatchSnapshot();
    });

    test('Should create  product', async () => {
        const store = setupStore();
        store.dispatch(addProducts(productsData));
        renderWithProviders( <MemoryRouter><CreateProduct  /></MemoryRouter>, {store})
        const initialLength = store.getState().products.products.length;
        expect(initialLength).toEqual(10);
        let nameInput = screen.getByPlaceholderText('Specify product name');
        expect(nameInput).toBeInTheDocument();
        fireEvent.change(nameInput, { target: { value: 'Test Product' } });
        expect(nameInput).toHaveValue('Test Product');
        let weightInput = screen.getByPlaceholderText('Weight');
        expect(weightInput).toBeInTheDocument();
        fireEvent.change(weightInput, { target: { value: '100' } });
        expect(weightInput).toHaveValue('100');
        let submitButton = screen.getByText('Add');
        fireEvent.click(
            submitButton,
            new MouseEvent('click', {
                bubbles: true,
                cancelable: true,
            })
        );
        let products = store.getState().products.products.length;
        expect(products).toBeGreaterThan(initialLength);
    });
})

