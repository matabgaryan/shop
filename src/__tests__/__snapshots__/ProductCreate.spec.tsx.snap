// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Products Take Snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="mt-16 mx-auto max-w-4xl w-full flex flex-col items-center"
    >
      <p
        class="mb-8 font-GoogleSansBold text-[18px] text-appBlack font-[700]"
      >
        Configure Product
      </p>
      <form>
        <div
          class="grid gap-4  mb-6 lg:grid-cols-3 mb-6 rounded-sm border-b border-lightGrey	pb-4"
        >
          <div
            class="mr-0 md:mr-4 undefined"
          >
            <label
              class="font-[400] pl-4 block mb-2 text-sm font-GoogleSansRegular text-black text-[16px]"
            >
              Name
            </label>
            <input
              class="w-full lg:w-64 placeholder-lightGrey  bg-transparent border border-oziLightGrey text-gray-900 text-sm rounded-md focus:ring-oziLightGrey focus:border-oziLightGrey p-2.5 "
              name="name"
              placeholder="Specify product name"
              type="text"
            />
          </div>
          <div
            class="mr-0 md:mr-4 undefined"
          >
            <label
              class="font-[400] pl-4 block mb-2 text-sm font-GoogleSansRegular text-black text-[16px]"
            >
              Price
            </label>
            <input
              class="w-full lg:w-64 placeholder-lightGrey  bg-transparent border border-oziLightGrey text-gray-900 text-sm rounded-md focus:ring-oziLightGrey focus:border-oziLightGrey p-2.5 "
              name="price"
              placeholder="Specify product price"
              type="text"
            />
          </div>
          <div
            class="mr-0 md:mr-4 undefined"
          >
            <label
              class="font-[400] pl-4 block mb-2 text-sm font-GoogleSansRegular text-black text-[16px]"
            >
              Weight
            </label>
            <input
              class="w-full lg:w-64 placeholder-lightGrey  bg-transparent border border-oziLightGrey text-gray-900 text-sm rounded-md focus:ring-oziLightGrey focus:border-oziLightGrey p-2.5 "
              name="weight"
              placeholder="Weight"
              type="text"
            />
          </div>
          <div
            class="mr-0 md:mr-4 undefined"
          >
            <label
              class="font-[400] pl-4 block mb-2 text-sm font-GoogleSansRegular text-black text-[16px]"
            >
              Start Date
            </label>
            <input
              class="w-full lg:w-64 placeholder-lightGrey  bg-transparent border border-oziLightGrey text-gray-900 text-sm rounded-md focus:ring-oziLightGrey focus:border-oziLightGrey p-2.5 "
              name="startDate"
              placeholder="Start Date"
              type="text"
            />
          </div>
          <div
            class="mr-0 md:mr-4 undefined"
          >
            <label
              class="font-[400] pl-4 block mb-2 text-sm font-GoogleSansRegular text-black text-[16px]"
            >
              End Date
            </label>
            <input
              class="w-full lg:w-64 placeholder-lightGrey  bg-transparent border border-oziLightGrey text-gray-900 text-sm rounded-md focus:ring-oziLightGrey focus:border-oziLightGrey p-2.5 "
              name="endDate"
              placeholder="End Date"
              type="text"
            />
          </div>
          <div
            class="flex flex-row  items-start justify-center"
          >
            <button
              class=" text-white bg-oziRed hover:opacity-70  font-medium rounded-lg text-sm  px-5 py-2.5 text-center "
            >
              Remove
            </button>
          </div>
        </div>
        <div>
          <div
            class="flex items-start justify-center"
          >
            <button
              class="text-appBlack text-[16px] font-GoogleSansRegular font-[400] focus:outline-none  text-center"
            >
              + Add more
            </button>
          </div>
        </div>
        <div
          class="flex justify-end mb-8 mt-5"
        >
          <button
            class="mr-4 w-150 text-appBlack bg-transparent hover:opacity-80	 focus:ring-4 font-medium rounded-lg text-sm  px-5 py-2.5 text-center "
          >
            Cancel
          </button>
          <button
            class="w-150 text-white bg-appBlack hover:opacity-70  font-medium rounded-lg text-sm  px-5 py-2.5 text-center "
            type="submit"
          >
            Add
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
